x-pack/plugins/fleet/server/routes/agent/handlers.ts

```diff
@@ -308,26 +308,32 @@ export const postBulkAgentsReassignHandler: RequestHandler<
 
   const soClient = context.core.savedObjects.client;
   const esClient = context.core.elasticsearch.client.asInternalUser;
+  const agentOptions = Array.isArray(request.body.agents)
+    ? { agentIds: request.body.agents }
+    : { kuery: request.body.agents };
 
   try {
     const results = await AgentService.reassignAgents(
       soClient,
       esClient,
-      Array.isArray(request.body.agents)
-        ? { agentIds: request.body.agents }
-        : { kuery: request.body.agents },
+      agentOptions,
       request.body.policy_id
     );
 
-    const body: PostBulkAgentReassignResponse = results.items.reduce((acc, so) => {
-      return {
-        ...acc,
-        [so.id]: {
-          success: !so.error,
-          error: so.error || undefined,
-        },
+    const body = results.items.reduce<PostBulkAgentReassignResponse>((acc, so) => {
+      acc[so.id] = {
+        success: !so.error,
+        error: so.error
+          ? {
+              name: so.error.name,
+              message: so.error.message,
+              // so.error.stack is also available
+            }
```

Comment:
what's the difference between this and just returning `so.error`?

Reply:
Thanks for the question. I meant to add a note about this. 

### TL; DR;
We need to be explicit because of how `JSON.stringify` deals with Error objects. We can standardize/generalize this later if we want.

### More info
When it was `so.error`, the response looked like `{ "name": "AgentReassignmentError" }`, instead of the message or stack trace I expected.

Looking further, I believe it's because the object is being given to `JSON.stringify`. deals with `Error` instances. For example,

```
> err = new Error('something')
Error: something
    at REPL21:1:7
    at Script.runInThisContext (vm.js:133:18)
    at REPLServer.defaultEval (repl.js:484:29)
    at bound (domain.js:413:15)
    at REPLServer.runBound [as eval] (domain.js:424:12)
    at REPLServer.onLine (repl.js:817:10)
    at REPLServer.emit (events.js:327:22)
    at REPLServer.EventEmitter.emit (domain.js:467:12)
    at REPLServer.Interface._onLine (readline.js:337:10)
    at REPLServer.Interface._line (readline.js:666:8)
> err.name
'Error'
> err.message
'something'
> err.stack
'Error: something\n' +
  '    at REPL21:1:7\n' +
  '    at Script.runInThisContext (vm.js:133:18)\n' +
  '    at REPLServer.defaultEval (repl.js:484:29)\n' +
  '    at bound (domain.js:413:15)\n' +
  '    at REPLServer.runBound [as eval] (domain.js:424:12)\n' +
  '    at REPLServer.onLine (repl.js:817:10)\n' +
  '    at REPLServer.emit (events.js:327:22)\n' +
  '    at REPLServer.EventEmitter.emit (domain.js:467:12)\n' +
  '    at REPLServer.Interface._onLine (readline.js:337:10)\n' +
  '    at REPLServer.Interface._line (readline.js:666:8)'
> JSON.stringify(err)
'{}'
```

Or using our specific example
```javascript
class IngestManagerError extends Error {
  constructor(message) {
    super(message);
    this.name = this.constructor.name; // for stack traces
  }
}
```

```
> fleetError = new IngestManagerError('oops')
IngestManagerError: oops
    at REPL73:1:14
    at Script.runInThisContext (vm.js:133:18)
    at REPLServer.defaultEval (repl.js:484:29)
    at bound (domain.js:413:15)
    at REPLServer.runBound [as eval] (domain.js:424:12)
    at REPLServer.onLine (repl.js:817:10)
    at REPLServer.emit (events.js:327:22)
    at REPLServer.EventEmitter.emit (domain.js:467:12)
    at REPLServer.Interface._onLine (readline.js:337:10)
    at REPLServer.Interface._line (readline.js:666:8)
```

```
> JSON.stringify(fleetError)
'{"name":"IngestManagerError"}'
```

### Some options
#### Define some other function(s) to control the formatting
  * Add them as a `toJSON` method to any of the `Error` classes. All `JSON.stringify` calls will call that as well.
  * Call them in a handler

Reply:
I see, thanks for the explanation. is it important to keep it as an object in the API response? if not, maybe we can consider returning [`.toString()`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/toString) instead?

Reply:
No, it's not important and I expected a string at first. I'll try calling & returning `so.error.toString()`. Thanks.

Reply:
Did this in https://github.com/elastic/kibana/pull/95024/commits/3f0381cbe6fdf37230b4aa2e91c619eeaf7dd6d5